server/src/main/java/org/opensearch/index/translog/transfer/TranslogCkpFilesTransferManager.java

```diff
@@ -0,0 +1,116 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.translog.transfer;
+
+import org.opensearch.action.LatchedActionListener;
+import org.opensearch.common.blobstore.BlobPath;
+import org.opensearch.common.blobstore.stream.write.WritePriority;
+import org.opensearch.common.util.concurrent.ConcurrentCollections;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.index.shard.ShardId;
+import org.opensearch.index.remote.RemoteTranslogTransferTracker;
+import org.opensearch.index.translog.transfer.FileSnapshot.CheckpointFileSnapshot;
+import org.opensearch.index.translog.transfer.FileSnapshot.TransferFileSnapshot;
+import org.opensearch.index.translog.transfer.FileSnapshot.TranslogFileSnapshot;
+import org.opensearch.indices.RemoteStoreSettings;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Translog transfer manager to transfer {@link TranslogCheckpointSnapshot} by transfering {@link TranslogFileSnapshot} and {@link CheckpointFileSnapshot} separately
+ *
+ * @opensearch.internal
+ */
+public class TranslogCkpFilesTransferManager extends TranslogTransferManager {
+
+    TransferService transferService;
+    FileTransferTracker fileTransferTracker;
+
+    public TranslogCkpFilesTransferManager(
+        ShardId shardId,
+        TransferService transferService,
+        BlobPath remoteDataTransferPath,
+        BlobPath remoteMetadataTransferPath,
+        FileTransferTracker fileTransferTracker,
+        RemoteTranslogTransferTracker remoteTranslogTransferTracker,
+        RemoteStoreSettings remoteStoreSettings
+    ) {
+        super(
+            shardId,
+            transferService,
+            remoteDataTransferPath,
+            remoteMetadataTransferPath,
+            fileTransferTracker,
+            remoteTranslogTransferTracker,
+            remoteStoreSettings
+        );
+        this.transferService = transferService;
+        this.fileTransferTracker = fileTransferTracker;
+    }
+
+    @Override
+    public void transferTranslogCheckpointSnapshot(
+        Set<TranslogCheckpointSnapshot> toUpload,
+        Map<Long, BlobPath> blobPathMap,
+        LatchedActionListener<TranslogCheckpointSnapshot> latchedActionListener
+    ) throws Exception {
+        for (TranslogCheckpointSnapshot tlogAndCkpTransferFileSnapshot : toUpload) {
+            Set<TransferFileSnapshot> filesToUpload = new HashSet<>();
+            Set<Exception> exceptionList = ConcurrentCollections.newConcurrentSet();
+
+            String tlogFileName = tlogAndCkpTransferFileSnapshot.getTranslogFileName();
+            String ckpFileName = tlogAndCkpTransferFileSnapshot.getCheckpointFileName();
+
+            if (!fileTransferTracker.uploaded(tlogFileName)) {
```

Comment:
`TranslogTransferManager` already has filtered files, right? Why do we need to filter them again?

Reply:
This class `TranslogCkpFilesTransferManager` transfer ckp and tlog files separately. `TranslogTransferManager` filtered out based on the `generation` but here in this class we are filtering out based on `fileName` as with this method if one of the files (lets say ckp uploaded but tlog didn't we mark the generation as not successful). In the retry we are filtering out files that are already uploaded.